lex/lexer.ml

```diff
@@ -0,0 +1,306 @@
+open Syntax
+open Parser
+
+exception Lexical_error of string * string * int * int
+exception Bad_rule
+
+let brace_depth = ref 0
+
+(* https://caml.inria.fr/pub/docs/manual-ocaml/libref/Lexing.html *)
+
+let rec get_next_buff lexbuf n =
+  (* Return the next {n} characters in the buffer *)
+  let p = lexbuf.Lexing.lex_curr_p in
+  let c_pos = p.Lexing.pos_cnum in
+  
+  (* if there is not enough characters in the buffer, fill it again *)
+  if lexbuf.Lexing.lex_buffer_len < n + c_pos
+  then begin
+    let curr_len = lexbuf.Lexing.lex_buffer_len in
+    lexbuf.Lexing.refill_buff lexbuf;
+    let new_len = lexbuf.Lexing.lex_buffer_len in
+    if curr_len != new_len
```

Comment:
`!=` is physical inequality and `==` is physical equality. When you don't need physical equality or inequality (for integers or booleans, for instance), it would be cleaner to use structural (in)equality with `<>` and `=`.

Reply:
Thank you! I changed my commit to have only structural equality tests.